contrib/format-msgpack/src/main/java/org/apache/drill/exec/store/msgpack/MsgpackReader.java

```diff
@@ -0,0 +1,192 @@
+  protected MessageUnpacker unpacker;
+  protected MsgpackReaderContext context;
+  private final boolean skipQuery;
+
+  /**
+   * Collection for tracking empty array writers during reading and storing them
+   * for initializing empty arrays
+   */
+  private final List<ListWriter> emptyArrayWriters = Lists.newArrayList();
+  private boolean allTextMode = false;
+  private MapValueWriter mapValueWriter;
+
+  public MsgpackReader(InputStream stream, MsgpackReaderContext context, DrillBuf managedBuf, List<SchemaPath> columns,
+      boolean skipQuery) {
+
+    this.context = context;
+    this.context.workBuf = managedBuf;
+    this.unpacker = MessagePack.newDefaultUnpacker(stream);
+    this.columns = columns;
+    this.skipQuery = skipQuery;
+    rootSelection = FieldSelection.getFieldSelection(columns);
+    EnumMap<ValueType, AbstractValueWriter> valueWriterMap = new EnumMap<>(ValueType.class);
+    valueWriterMap.put(ValueType.ARRAY, new ArrayValueWriter(valueWriterMap, emptyArrayWriters));
+    valueWriterMap.put(ValueType.FLOAT, new FloatValueWriter());
+    valueWriterMap.put(ValueType.INTEGER, new IntegerValueWriter());
+    valueWriterMap.put(ValueType.BOOLEAN, new BooleanValueWriter());
+    valueWriterMap.put(ValueType.STRING, new StringValueWriter());
+    valueWriterMap.put(ValueType.BINARY, new BinaryValueWriter());
+    valueWriterMap.put(ValueType.EXTENSION, new ExtensionValueWriter());
+    mapValueWriter = new MapValueWriter(valueWriterMap);
+    valueWriterMap.put(ValueType.MAP, mapValueWriter);
+
+    for (AbstractValueWriter w : valueWriterMap.values()) {
+      w.setup(this.context);
+    }
+  }
+
+  public boolean write(ComplexWriter writer, MaterializedField schema)
+      throws IOException, MessageInsufficientBufferException {
+    if (!unpacker.hasNext()) {
+      return false;
+    }
+
+    context.useSchema = schema != null;
+
+    try {
+      Value v = unpacker.unpackValue();
+      ValueType type = v.getValueType();
+      if (type == ValueType.MAP) {
+        writeRecord(v.asMapValue(), writer, schema);
+      } else {
+        throw new MsgpackParsingException(
+            "Value in root of message pack file is not of type MAP. Skipping type found: " + type);
+      }
+    } catch (MessageInsufficientBufferException e) {
+      throw new MsgpackParsingException("Failed to unpack MAP, possibly because key/value tuples do not match.");
+    }
+    return true;
+  }
+
+  protected void writeRecord(MapValue value, ComplexWriter writer, MaterializedField schema) throws IOException {
+    if (skipQuery) {
+      writer.rootAsMap().bit("count").writeBit(1);
+    } else {
+      mapValueWriter.writeToMap(value, writer.rootAsMap(), this.rootSelection, schema);
+    }
+  }
+
+//
+//
+//  private void ensure(final int length) {
+//    workBuf = workBuf.reallocIfNeeded(length);
+//  }
+
+  @SuppressWarnings("resource")
+  public void ensureAtLeastOneField(ComplexWriter writer) {
```

Comment:
Looks like you happened to look at one of the more complex readers for an example. The example cited above shows a much easier way to do this.